optaplanner-core/src/main/java/org/optaplanner/core/api/score/stream/ConstraintCollectors.java

```diff
@@ -44,107 +46,82 @@
-        return new DefaultBiConstraintCollector<>(
-                () -> new int[1],
-                (resultContainer, a, b) -> {
-                    resultContainer[0]++;
-                    return (() -> resultContainer[0]--);
-                },
-                resultContainer -> resultContainer[0]);
+        return countLongBi(Long::intValue);
     }
 
     public static <A, B> BiConstraintCollector<A, B, ?, Long> countLongBi() {
+        return countLongBi(Functions.identity());
+    }
+
+    private static <A, B, C extends Number> BiConstraintCollector<A, B, ?, C> countLongBi(Function<Long, C> countConverter) {
         return new DefaultBiConstraintCollector<>(
-                () -> new long[1],
+                LongAdder::new,
                 (resultContainer, a, b) -> {
-                    resultContainer[0]++;
-                    return (() -> resultContainer[0]--);
+                    resultContainer.increment();
+                    return (resultContainer::decrement);
                 },
-                resultContainer -> resultContainer[0]);
+                resultContainer-> countConverter.apply(resultContainer.longValue()));
     }
 
     // ************************************************************************
     // countDistinct
     // ************************************************************************
 
     public static <A> UniConstraintCollector<A, ?, Integer> countDistinct(Function<A, ?> groupValueMapping) {
-        class CountDistinctResultContainer {
-            int count = 0;
-            Map<Object, int[]> objectCountMap = new HashMap<>();
-        }
-        return new DefaultUniConstraintCollector<>(
-                CountDistinctResultContainer::new,
-                (resultContainer, a) -> {
-                    Object value = groupValueMapping.apply(a);
-                    int[] objectCount = resultContainer.objectCountMap.computeIfAbsent(value, k -> new int[1]);
-                    if (objectCount[0] == 0) {
-                        resultContainer.count++;
-                    }
-                    objectCount[0]++;
-                    return (() -> {
-                        int[] objectCount2 = resultContainer.objectCountMap.get(value);
-                        if (objectCount2 == null) {
-                            throw new IllegalStateException("Impossible state: the value (" + value
-                                    + ") of A (" + a + ") is removed more times than it was added.");
-                        }
-                        objectCount2[0]--;
-                        if (objectCount2[0] == 0) {
-                            resultContainer.objectCountMap.remove(value);
-                            resultContainer.count--;
-                        }
-                    });
-                },
-                resultContainer -> resultContainer.count);
+        return ConstraintCollectors.countDistinctLong(groupValueMapping, Long::intValue);
```

Comment:
cast back and forth to longs might impact performance
but definitely impacts behavior when overflowing:
2 000 000 000 + 1 000 000 000 will result in a postive number after this change, instead of a negative number (which is easier to spot for overflow)